text/0000-cargo-binary-dependencies.md

```diff
@@ -0,0 +1,156 @@
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin" }
+```
+
+Cargo will build the `cmake` binary, then make it available to your `build.rs` through an environment variable:
+
+```rust
+// build.rs
+use std::{env, process::Command};
+
+fn main() {
+    let cmake_path = env::var_os("CARGO_BIN_FILE_CMAKE_cmake").expect("cmake binary");
+    let mut cmake = Command::new(cmake_path).arg("--version");
+    assert!(cmake.status().expect("cmake --version failed").success());
+}
+```
+
+You can optionally specify specific binaries to depend on using `bins`:
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin", bins = ["cmake"] }
+```
+
+If no binaries are specified, all the binaries in the package will be built and made available.
+
+You can obtain the directory containing all binaries built by the `cmake` crate with `CARGO_BIN_DIR_CMAKE`, such as to add it to `$PATH` before invoking another build system or a script.
+
+Cargo also allows depending on `cdylib` or `staticlib` artifacts. For example, you can embed a dynamic library in your binary:
+
+```rust
+// main.rs
+const MY_PRELOAD_LIB: &[u8] = include_bytes!(env!("CARGO_CDYLIB_FILE_MYPRELOAD"));
+```
+
+Note that cargo cannot help you ensure these artifacts are available at runtime for an installed version of a binary; cargo can only supply these artifacts at build time. Runtime requirements for installed crates are out of scope for this change.
+
+If you need to depend on multiple variants of a crate, such as both the binary and library of a crate, you can supply an array of strings for `type`: `type = ["bin", "lib"]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are four `type`s available:
+1. `"lib"`, the default
+2. `"bin"`, a crate building one or more binaries
+3. `"cdylib"`, a C-compatible dynamic library
+4. `"staticlib"`, a C-compatible static library
+
+`"lib"` corresponds to all crates that can be depended on currently,
+including `lib`, `rlib`, and `proc-macro` libraries.
+See [linkage](https://doc.rust-lang.org/reference/linkage.html) for more information.
+
+Artifact dependencies can appear in any of the three sections of dependencies (or in target-specific versions of these sections):
+- `[build-dependencies]`
+- `[dependencies]`
+- `[dev-dependencies]`
+
+By default, `build-dependencies` are built for the host, while  `dependencies` and `dev-dependencies` are built for the target. You can specify the `target` attribute to build for a specific target, such as `target = "wasm32-wasi"`; a literal `target = "target"` will build for the target even if specifing a build dependency. (If the target is not available, this will result in an error at build time, just as if building the specified crate with a `--target` option for an unavailable target.)
```

Comment:
> How does the `target` field work if Cargo is building for multiple targets? It seems like the environment variables are designed to assume only one target is built.

If you want to build an artifact dependency for multiple targets, you need to specify it multiple times with different names.

```toml
[dependencies]
vdso32 = { package = "my-vdso", version = "1.2.3", target = "i686-unknown-linux-gnu", artifact = "cdylib" }
vdso64 = { package = "my-vdso", version = "1.2.3", target = "x86_64-unknown-linux-gnu", artifact = "cdylib" }
```

You'll then have `CARGO_CDYLIB_FILE_VDSO32` and `CARGO_CDYLIB_FILE_VDSO64`.

> Also, as an aside, `target="target"` seems like it could be a little confusing or unclear. I don't have any particularly better suggestions, though.

I felt the same way about it. I think the terminology of "host" vs "target" is sufficiently clear to make it relatively reasonable. As an alternative, would you find `target = true` less confusing, or more confusing?

Reply:
What I was concerned about is that if you specify `target = "target"`, what does `cargo build --target x86_64-unknown-linux-gnu --target i686-unknown-linux-gnu` do?

Reply:
@ehuss Exactly the same as it does with any other dependency that needs to be built for the target: built it for each target. If crate `abc` has an artifact dependency on crate `xyz` with `target="target"`, then the `x86_64-unknown-linux-gnu` build of `abc` requires an `x86_64-unknown-linux-gnu` build of `xyz`, and the `i686-unknown-linux-gnu` build of `abc` requires an `i686-unknown-linux-gnu` build of `xyz`. Each of those builds will get the corresponding artifact dependency they require.

Reply:
@ehuss I've added a sentence to this effect in the RFC.